---
title: "Analysis features positive"
output: pdf_document
date: '2022-06-14'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 80), tidy = TRUE)
```


```{r libraries}
library(factoextra)
library(corrplot)
library(FactoMineR)
library(NMF)
library(RColorBrewer)
```

```{r load_data}
load("/Users/carlacasanovasuarez/Documents/Master Bioinformatics UAB/Prácticas Radiomics/Radiomic features/Results_rfeatures/R objects/rdr_assay_scaled.rda")
load("/Users/carlacasanovasuarez/Documents/Master Bioinformatics UAB/Prácticas Radiomics/Radiomic features/Results_rfeatures/R objects/sputum_eset_countsOK.rda")
load("/Users/carlacasanovasuarez/Documents/Master Bioinformatics UAB/Prácticas Radiomics/Radiomic features/Results_rfeatures/R objects/sputum_eset_phenoOK.rda")
```


```{r check_assay}
head(rdr_assay, 3)
```


## Corralation analysis: features with positive DE  

```{r positive_features}
pos_rf <- list("Sphericity.original",
               "Variance.original",
               "Autocorrelation.original",
               "ClusterShade.original",
               "Contrast.original",
               "DifferenceAverage.original",
               "Id.original",
               "Idm.original",
               "Idmn.original",
               "Idn.original",
               "InverseVariance.original",
               "JointAverage.original",
               "SumAverage.original",
               "HighGrayLevelEmphasis.original",
               "LongRunLowGrayLevelEmphasis.original",
               "LowGrayLevelRunEmphasis.original",
               "RunPercentage.original",
               "LargeAreaEmphasis.original",
               "LowGrayLevelEmphasis.original",
               "LargeAreaHighGrayLevelEmphasis.original",
               "LargeAreaLowGrayLevelEmphasis.original",
               "SmallAreaEmphasis.original",
               "SmallAreaHighGrayLevelEmphasis.original",
               "SmallAreaLowGrayLevelEmphasis.original",
               "ZoneVariance.original",
               "DependenceNonUniformity.original",
               "HighGrayLevelRunEmphasis.original",
               "LargeDependenceEmphasis.original",
               "LargeDependenceLowGrayLevelEmphasis.original",
               "Busyness.original",
               "Coarseness.original",
               "Complexity.original",
               "Strength.original",
               "X10Percentile.original",
               "InterquartileRange.original")
```


Subset original data of patients with transcriptomics:  

```{r subset}
# Prepare data in object: rdr_assay.scaled
rownames(rdr_assay)[rownames(rdr_assay) == "10Percentile.original"] <- "X10Percentile.original"
rownames(rdr_assay)[rownames(rdr_assay) == "90Percentile.original"] <- "X90Percentile.original"

# Store an array with values of features associated to genes DE
rdr_positive <- rdr_assay[unlist(pos_rf),]

# Change labels to visualize better
rownames(rdr_positive) <- substr(rownames(rdr_positive), start = 1, stop =  nchar(rownames(rdr_positive))-9)
```

See correlations between features positively associated to genes DE:  

```{r corr_matrix}
cor.mat <- round(cor(t(rdr_positive)),2)

head(cor.mat,4)
```


```{r corrplot}
#pdf("Radiomic features from DE analysis correlation matrix.pdf")

corrplot(cor.mat, type="upper", order="hclust", 
         tl.col="black", tl.srt=45, tl.cex = 0.4)

#dev.off()
```


## PCA  

```{r pca_all_transcriptomics}
# Change labels to visualize better
rownames(rdr_assay) <- substr(rownames(rdr_assay), start = 1, stop =  nchar(rownames(rdr_assay))-9)

# Perform PCA for all features, but only 125 patients
res.pca <- PCA(t(rdr_assay), graph = FALSE)
```

### Variables contribution  

```{r results_var}
# Extract the results for variables
var <- get_pca_var(res.pca)
var
```

```{r var_results_explore}
head(var$cor)
```


```{r var}
#pdf("Contribution variables PCA 125 patients.pdf")

# Graph of variables: default plot
fviz_pca_var(res.pca, col.var="contrib",
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE) # Avoid text overlapping

#dev.off()
```


```{r explain_variability}
fviz_screeplot(res.pca, ncp=10)
```


```{r contrib_var}
#pdf("Contribution variables PC1-3.pdf")

# Contributions of variables to PC1
fviz_contrib(res.pca, choice = "var", axes = 1, top = 10)
# Contributions of variables to PC2
fviz_contrib(res.pca, choice = "var", axes = 2, top = 10)
# Contributions of variables to PC3
fviz_contrib(res.pca, choice = "var", axes = 3, top = 10)

#dev.off()
```

### Individuals

```{r individuals_graph}
#pdf("Contribution individuals PCA 125 patients.pdf")

# Graph of individuals
# 1. Use repel = TRUE to avoid overplotting
# 2. Control automatically the color of individuals using the cos2
    # cos2 = the quality of the individuals on the factor map
    # Use points only
# 3. Use gradient color
fviz_pca_ind(res.pca, col.ind = "cos2",
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE # Avoid text overlapping (slow if many points)
             )

#dev.off()
```


```{r biplot_indv_var}
# Biplot of individuals and variables
fviz_pca_biplot(res.pca, repel = TRUE)
```


```{r indv}
#pdf("Contribution individuals by group.pdf")

# Visualize
# Use habillage to specify groups for coloring
fviz_pca_ind(res.pca,
             label = "none", # hide individual labels
             habillage = phenoDataFrame$GROUP_bin, # color by groups
             palette = c("#00AFBB", "#E7B800"),
             addEllipses = TRUE # Concentration ellipses
             )

#dev.off()
```

## HCL  

Variables must be columns and observations row:  

```{r heatmap}
#pdf("Heatmap positively correlated features with genes DE.pdf")

# Colors
mypalette <- brewer.pal(11, "PiYG")
morecols <- colorRampPalette(mypalette)

# Heatmap
aheatmap(rdr_positive, col=rev(morecols(37)), main = "Rfeatures with genes DE", scale = "none")

#dev.off()
```


```{r asses_k_clusters}
set.seed(123)

fviz_nbclust(rdr_positive, kmeans, method = "silhouette", k.max = 20) + 
  labs(subtitle = "Silhouette method")

fviz_nbclust(rdr_positive, kmeans, method = "wss", k.max = 20) + 
  geom_vline(xintercept = 8, linetype = 2) +
  labs(subtitle = "Elbow method")
```

**Gap statistics** measures how different the total within intra-cluster variation can be between observed data and reference data with a **random uniform distribution**. A large gap statistics means the clustering structure is very far away from the random uniform distribution of points.
The number of clusters can be chosen as the smallest value of k such that the gap statistic is within one **standard deviation** of the gap at k+1. In your case, when k = 12, its value is greater than the value that k = 13 minus one standard deviation.  

`maxSE(f, SE.f)` is included as an argument of `fviz_gap_stat()` (unlike `fivz_nbclust`) and it determines the location of the maximum of $f$, taking a "1-SE rule" into account for the *SE* methods. The default method `firstSEmax` looks for the smallest k such that its value $f(k)$ is not more than 1 standard error away from the first local maximum. 

```{r test}
set.seed(123)

gap_stat <- clusGap(rdr_positive, FUN = kmeans, nstart = 25, K.max = 30, B = 100)
fviz_gap_stat(gap_stat) + theme_minimal() + ggtitle("fviz_gap_stat: Gap Statistic")
```



```{r hcl}
#pdf("Dendogram positive features.pdf")

# Compute hierarchical clustering and cut into 4 clusters
res <- hcut(rdr_positive, k = 17, stand = TRUE)
# Visualize
fviz_dend(res, rect = TRUE, cex = 0.4, lwd = 0.5, labels_track_height = 20,
          k_colors = c("#00AFBB","#2E9FDF", "#E7B800", "#FC4E07"))

#dev.off()
```


```{r }
k.res <- kmeans(rdr_positive, centers = 17, nstart = 25)

fviz_cluster(k.res, 
             geom = "text", data = rdr_positive,
             ggtheme = theme_minimal(),
             main = "Partitioning Clustering Plot",
             repel = TRUE)
```







